addons/threadsafety.py

```diff
@@ -1,39 +1,401 @@
     for var in data.variables:
         if var.isStatic and var.isLocal:
-            type = None
+            vartype = None
             if var.isClass:
-                type = 'object'
+                vartype = 'object'
             else:
-                type = 'variable'
+                vartype = 'variable'
             if var.isConst:
                 if data.standards.cpp == 'c++03':
-                    reportError(var.typeStartToken, 'warning', 'Local constant static ' + type + ' \'' + var.nameToken.str + '\', dangerous if it is initialized in parallel threads', 'threadsafety-const')
+                    reportError(
+                        var.typeStartToken,
+                        'warning',
+                        'Local constant static '
+                        + vartype + "'" + var.nameToken.str
+                        + "', dangerous if it is initialized"
+                        + ' in parallel threads',
+                        'threadsafety-const')
             else:
-                reportError(var.typeStartToken, 'warning', 'Local static ' + type + ': ' + var.nameToken.str, 'threadsafety')
+                reportError(var.typeStartToken, 'warning',
+                            'Local static ' + vartype + ': '
+                            + var.nameToken.str,
+                            'threadsafety')
 
-for arg in sys.argv[1:]:
-    if arg.startswith('-'):
-        continue
 
-    print('Checking %s...' % arg)
-    data = cppcheckdata.CppcheckData(arg)
+def check_MTunsafe(data, srcfile, quiet=False):
+    """
+    Look for functions marked MT-unsafe in their man pages.
 
+    The MT-unsafe functions are listed in id_MTunsafe (and id_MTunsafe_full).
+    That section of code can be regenerated by the external script MT-Unsafe.py
+    """
+    # Assume that the code is Multi-thread safe until proven otherwise
+    MTsafe = True
+
+    # go through each configuration
     for cfg in data.iterconfigurations():
-        print('Checking %s, config %s...' % (arg, cfg.name))
-        checkstatic(cfg)
+        if not quiet:
+            print('Checking %s, config %s...' % (srcfile, cfg.name))
+        safe_ranges = []
+        safe = -1
+
+        # check end of source beyond last directive
+        if len(cfg.tokenlist) > 0:
+            unsafe = int(cfg.tokenlist[-1].linenr)
+            if unsafe > safe > 0:
+                safe_ranges.append((safe, unsafe))
+
+        # go through all tokens
+        for token in cfg.tokenlist:
+            if token.str in id_MTunsafe:
+                if not any(lower <= int(token.linenr) <= upper
+                           for (lower, upper) in safe_ranges):
+                    cppcheckdata.reportError(token, 'warning',
+                                             token.str + ' is MT-unsafe',
+                                             'threadsafety',
+                                             'unsafe-call')
+                    MTsafe = False
+            token = token.next
```

Comment:
Sounds good.
It may be a week or so before I can work on this again.

Not sure what I did to close this PR. I definitely did not mean to do so.
Can I unclose it or do I need to make another PR ?